<a href="https://colab.research.google.com/github/hamasli/Email-Spam-Classification/blob/main/Email_Spam_Classification.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [2]:
!pip install pandas numpy scikit-learn


In [17]:
import pandas as pd;
import numpy as np;
import sklearn as sk
from sklearn.model_selection import train_test_split;

In [5]:
df=pd.read_csv('emails.csv');
df

Email No.  the  to  ect  and  for  of    a  you  hou  ...  connevey  \
0        Email 1    0   0    1    0    0   0    2    0    0  ...       0.0   
1        Email 2    8  13   24    6    6   2  102    1   27  ...       0.0   
2        Email 3    0   0    1    0    0   0    8    0    0  ...       0.0   
3        Email 4    0   5   22    0    5   1   51    2   10  ...       0.0   
4        Email 5    7   6   17    1    5   2   57    0    9  ...       0.0   
...          ...  ...  ..  ...  ...  ...  ..  ...  ...  ...  ...       ...   
4335  Email 4336    0   1    1    0    0   1    3    1    0  ...       0.0   
4336  Email 4337   40  27    2   33    5  18  242    9    3  ...       0.0   
4337  Email 4338    0   0    1    0    1   0    3    0    0  ...       0.0   
4338  Email 4339    2   3    5    3    4   5   35    3    0  ...       0.0   
4339  Email 4340   47  32   11   29   17  22  280    5    2  ...       NaN   

      jay  valued  lay  infrastructure  military  allowing   ff  dry  \
0     0.0     0.0  0.0             0.0       0.0       0.0  0.0  0.0   
1     0.0     0.0  0.0             0.0       0.0       0.0  1.0  0.0   
2     0.0     0.0  0.0             0.0       0.0       0.0  0.0  0.0   
3     0.0     0.0  0.0             0.0       0.0       0.0  0.0  0.0   
4     0.0     0.0  0.0             0.0       0.0       0.0  1.0  0.0   
...   ...     ...  ...             ...       ...       ...  ...  ...   
4335  0.0     0.0  0.0             0.0       0.0       0.0  0.0  0.0   
4336  0.0     0.0  1.0             0.0       0.0       0.0  8.0  0.0   
4337  0.0     0.0  0.0             0.0       0.0       0.0  0.0  0.0   
4338  0.0     0.0  0.0             0.0       0.0       0.0  4.0  0.0   
4339  NaN     NaN  NaN             NaN       NaN       NaN  NaN  NaN   

      Prediction  
0            0.0  
1            0.0  
2            0.0  
3            0.0  
4            0.0  
...          ...  
4335         0.0  
4336         0.0  
4337         0.0  
4338         1.0  
4339         NaN  

[4340 rows x 3002 columns]

**After loading the dataset now we will apply preprocessing on the dataset**

In [6]:
print(df.info());

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 4340 entries, 0 to 4339
Columns: 3002 entries, Email No. to Prediction
dtypes: float64(593), int64(2408), object(1)
memory usage: 99.4+ MB
None


In [8]:
column_names=df.columns.tolist();
column_names

['Email No.',
 'the',
 'to',
 'ect',
 'and',
 'for',
 'of',
 'a',
 'you',
 'hou',
 'in',
 'on',
 'is',
 'this',
 'enron',
 'i',
 'be',
 'that',
 'will',
 'have',
 'with',
 'your',
 'at',
 'we',
 's',
 'are',
 'it',
 'by',
 'com',
 'as',
 'from',
 'gas',
 'or',
 'not',
 'me',
 'deal',
 'if',
 'meter',
 'hpl',
 'please',
 're',
 'e',
 'any',
 'our',
 'corp',
 'can',
 'd',
 'all',
 'has',
 'was',
 'know',
 'need',
 'an',
 'forwarded',
 'new',
 't',
 'may',
 'up',
 'j',
 'mmbtu',
 'should',
 'do',
 'am',
 'get',
 'out',
 'see',
 'no',
 'there',
 'price',
 'daren',
 'but',
 'been',
 'company',
 'l',
 'these',
 'let',
 'so',
 'would',
 'm',
 'into',
 'xls',
 'farmer',
 'attached',
 'us',
 'information',
 'they',
 'message',
 'day',
 'time',
 'my',
 'one',
 'what',
 'only',
 'http',
 'th',
 'volume',
 'mail',
 'contract',
 'which',
 'month',
 'more',
 'robert',
 'sitara',
 'about',
 'texas',
 'nom',
 'energy',
 'pec',
 'questions',
 'www',
 'deals',
 'volumes',
 'pm',
 'ena',
 'now',
 'their'

In [9]:
print(len(column_names))

3002


In [10]:
df=df.drop(columns=df.columns[0]);
x=df.drop(columns=['Prediction']);
y=df['Prediction'];
print(y);
print(df)

0       0.0
1       0.0
2       0.0
3       0.0
4       0.0
       ... 
4335    0.0
4336    0.0
4337    0.0
4338    1.0
4339    NaN
Name: Prediction, Length: 4340, dtype: float64
      the  to  ect  and  for  of    a  you  hou   in  ...  connevey  jay  \
0       0   0    1    0    0   0    2    0    0    0  ...       0.0  0.0   
1       8  13   24    6    6   2  102    1   27   18  ...       0.0  0.0   
2       0   0    1    0    0   0    8    0    0    4  ...       0.0  0.0   
3       0   5   22    0    5   1   51    2   10    1  ...       0.0  0.0   
4       7   6   17    1    5   2   57    0    9    3  ...       0.0  0.0   
...   ...  ..  ...  ...  ...  ..  ...  ...  ...  ...  ...       ...  ...   
4335    0   1    1    0    0   1    3    1    0    1  ...       0.0  0.0   
4336   40  27    2   33    5  18  242    9    3   41  ...       0.0  0.0   
4337    0   0    1    0    1   0    3    0    0    0  ...       0.0  0.0   
4338    2   3    5    3    4   5   35    3    0    3  ...    

In [14]:
print(x.isnull().sum().sum());

592


In [16]:
#here will will write 0 at all the places where column row values are null
x=x.fillna(0);
print(x.isnull().sum().sum());


0


In [37]:
y=y.fillna(0);
print(y.isnull().sum());

0


In [38]:
# df.shape
print(df['Prediction'].value_counts())

Prediction
0.0    3100
1.0    1239
Name: count, dtype: int64


now you can see that there is no missing value in dataset

In [39]:
x_train,x_test,y_train,y_test=train_test_split(x,y,test_size=0.3,random_state=42);
print("Training set size:", x_train.shape[0])
print("Testing set size:", x_test.shape[0])

Training set size: 3038
Testing set size: 1302


In [40]:
from sklearn.feature_selection import SelectKBest, chi2

selector = SelectKBest(chi2, k=1000)  # Choose the top 1000 features
X_train_selected = selector.fit_transform(x_train, y_train)
X_test_selected = selector.transform(x_test)

print("Selected features shape:", X_train_selected.shape)

Selected features shape: (3038, 1000)


You can see that we have performed the feature selection on our dataset and we have selected only top 1000 features

**Now we will build models and train them on training dataset.**

In [41]:
# Initialize Naive Bayes Multinomial model
from sklearn.naive_bayes import MultinomialNB
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report

nb_multinomial = MultinomialNB()
nb_multinomial.fit(X_train_selected, y_train)


MultinomialNB()

In [42]:

# Initialize Naive Bayes Gaussian model
from sklearn.naive_bayes import GaussianNB


nb_gaussian = GaussianNB()
nb_gaussian.fit(X_train_selected, y_train)


GaussianNB()

In [43]:
from sklearn.tree import DecisionTreeClassifier

# Initialize Decision Tree (J48) model
j48 = DecisionTreeClassifier()
j48.fit(X_train_selected, y_train)


DecisionTreeClassifier()

In [44]:
# Predict using Naive Bayes Multinomial model
y_pred_nb_multinomial = nb_multinomial.predict(X_test_selected)

print("Naive Bayes Multinomial Results:")
print("Accuracy:", accuracy_score(y_test, y_pred_nb_multinomial))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_nb_multinomial))
print("Classification Report:\n", classification_report(y_test, y_pred_nb_multinomial))


Naive Bayes Multinomial Results:
Accuracy: 0.946236559139785
Confusion Matrix:
 [[886  54]
 [ 16 346]]
Classification Report:
               precision    recall  f1-score   support

         0.0       0.98      0.94      0.96       940
         1.0       0.86      0.96      0.91       362

    accuracy                           0.95      1302
   macro avg       0.92      0.95      0.94      1302
weighted avg       0.95      0.95      0.95      1302



In [46]:
# Predict using Naive Bayes Gaussian model
y_pred_nb_gaussian = nb_gaussian.predict(X_test_selected)

# Evaluate the model
print("Naive Bayes Gaussian Results:")
print("Accuracy:", accuracy_score(y_test, y_pred_nb_gaussian))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_nb_gaussian))
print("Classification Report:\n", classification_report(y_test, y_pred_nb_gaussian))


Naive Bayes Gaussian Results:
Accuracy: 0.9032258064516129
Confusion Matrix:
 [[914  26]
 [100 262]]
Classification Report:
               precision    recall  f1-score   support

         0.0       0.90      0.97      0.94       940
         1.0       0.91      0.72      0.81       362

    accuracy                           0.90      1302
   macro avg       0.91      0.85      0.87      1302
weighted avg       0.90      0.90      0.90      1302



In [47]:
# Predict using Decision Tree (J48) model
y_pred_j48 = j48.predict(X_test_selected)

# Evaluate the model
print("Decision Tree (J48) Results:")
print("Accuracy:", accuracy_score(y_test, y_pred_j48))
print("Confusion Matrix:\n", confusion_matrix(y_test, y_pred_j48))
print("Classification Report:\n", classification_report(y_test, y_pred_j48))


Decision Tree (J48) Results:
Accuracy: 0.9308755760368663
Confusion Matrix:
 [[891  49]
 [ 41 321]]
Classification Report:
               precision    recall  f1-score   support

         0.0       0.96      0.95      0.95       940
         1.0       0.87      0.89      0.88       362

    accuracy                           0.93      1302
   macro avg       0.91      0.92      0.91      1302
weighted avg       0.93      0.93      0.93      1302



*You can see the performace of models on unseen(training data). We got a really god accuracy on all models greater than 0.90. And the best fit model on our dataset is Naive Bayes Multinomial model with the accuracy of 0.94*



